docs/ReferenceGuides/UnderscoredAttributes.md

```diff
@@ -0,0 +1,328 @@
+}
+```
+
+The temporary pointer conversion produces a pointer that is only
+guaranteed to be valid for the duration of the call to the initializer,
+and becomes invalid once the call ends.
+So the newly created `UnsafePointer` will be dangling.
+
+One exception to this is that inout-to-pointer conversions
+on static stored properties and global stored properties
+produce non-ephemeral pointers, as long as they have no observers:
+
+```swift
+var global = 0
+
+struct S {
+  static var staticVar = 0
+}
+
+func baz() {
+  let ptr = UnsafePointer(&global) // okay
+  let ptr2 = UnsafePointer(&S.staticVar) // okay
+}
+```
+
+Additionally, if they are of a tuple or struct type, their stored members
+without observers may also be passed inout as non-ephemeral pointers.
+
+For more details, see [the educational note on temporary pointer usage](/userdocs/diagnostics/temporary-pointers.md).
+
+## `@_nonoverride`
+
+## `@_objc_non_lazy_realization`
+
+## `@_optimize(...)`
+
+## `@_originallyDefinedIn`
+
+Marks a declaration as being originally defined in a different module,
+changing the name mangling. This can be used to move declarations
+from a module to one of the modules it imports without breaking clients.
+
+Consider the following example where a framework ToasterKit needs
+to move some APIs to a lower-level framework ToasterKitCore.
+Here are the necessary changes:
+
+1. Add a linker flag `-reexport_framework ToasterKitCore` for ToasterKit.
+   This ensures all symbols defined in ToasterKitCore will be accessible during
+   runtime via ToasterKit, so existing apps continue to run.
+2. In ToasterKit, use `@_exported import ToasterKitCore`.
+   This ensures existing source code that only imports ToasterKit continues to
+   type-check.
+3. Move the necessary declarations from ToasterKit to ToasterKitCore.
+   The moved declaration should have two attributes:
+   - `@available` indicating when the declaration was introduced in ToasterKit.
+   - `@_originallyDefinedIn` indicating the original module and when the
+     declaration was moved to ToasterKitCore.
+   ```
+   @available(toasterOS 42, *)
+   @_originallyDefinedIn(module: "ToasterKit", toasterOS 57)
+   enum Toast {
+   case underdone
+   case perfect
+   case burnt
+   }
+   ```
+4. Add Swift compiler flags `-Xfrontend -emit-ldadd-cfile-path -Xfrontend /tmp/t.c`
+   to ToasterKitCore's build settings. Add the emitted `/tmp/t.c` file to
+   ToasterKit's compilation.
+   This ensures when an app is built for deployment targets prior to the symbols' move,
+   the app will look for these symbols in ToasterKit instead of ToasterKitCore.
+
+## `@_private(sourceFile:)`
+
+Fully bypasses access control, allowing access to private declarations
+in the imported module. The imported module needs to be compiled with
+`-Xfrontend -enable-private-imports` for this to work.
+
+## `@_semantics`
+
+Marks a function as having particular high-level semantics that are
+specially recognized by the SIL optimizer.
```

Comment:
This is allows to optimizer to make use of some key invariants in performance critical data types, particularly Array. Since the implementation of those data types is written in Swift using Unsafe APIs, the optimizer would otherwise need to make conservative assumptions. Undefined behavior will result if the implementation changes in a way that violates the optimizer's assumptions about those semantics.